{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6ccbd969-ea19-4637-89f6-fa40587ad880",
   "metadata": {},
   "source": [
    "# Figure 1. common variant results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f39036d-5ac9-48af-9e83-46d068ef5000",
   "metadata": {},
   "source": [
    "Figure 1 combines the common variant results from:\n",
    "- MOSTEST (A+B)\n",
    "- Brainspan (from FUMA -> MAGMA) (C+D)\n",
    "- Betas from mostest for the 3 lead SNPs (E+F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a89e537d-b47c-466a-b282-b6f716565533",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import glob\n",
    "import h5py\n",
    "import scipy.io as sio\n",
    "import scipy.stats as stats\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "import seaborn as sns\n",
    "from nilearn import plotting, image\n",
    "from matplotlib import cm\n",
    "from matplotlib.colors import ListedColormap, LinearSegmentedColormap\n",
    "from matplotlib import colors as mcolors, path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73dea061-02bc-44df-a8f8-a8fe5bcb7be6",
   "metadata": {},
   "source": [
    "## Data loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9dad0b06-3cba-4fcc-8f47-f42c67f13b10",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def add_row(df, new_data, column_names, index_name, np_array=True):\n",
    "    if np_array:\n",
    "        df_new = pd.DataFrame(data=new_data.reshape((1, len(column_names))), columns=column_names, index=[index_name])\n",
    "    else:\n",
    "        df_new = pd.DataFrame.from_dict(new_data)\n",
    "    return pd.concat([df, df_new], axis=0)\n",
    "        \n",
    "def get_snp(chr_no, snp_id):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    bim_fn = \"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/mostest/in/mostest_geno_in_c{}.bim\".format(chr_no)\n",
    "    bim = pd.read_csv(bim_fn, sep='\\t', header=None, names='CHR SNP GP BP A1 A2'.split())\n",
    "    snp = bim.SNP.values == snp_id\n",
    "    return snp, bim.loc[bim.SNP.values == snp_id, :].to_dict()\n",
    "\n",
    "def get_file_name_mat(cat, chr_no):\n",
    "    if cat == \"edges\":\n",
    "        return \"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/mostest/out/sent_edges/edges_chr{}_zmat.mat\".format(chr_no)\n",
    "    if cat == \"edges_HD\":\n",
    "        return \"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/mostest/out/sent_edges_asym/edges_asym_chr{}_zmat.mat\".format(chr_no)\n",
    "\n",
    "def get_betas_pvals(fname, snps):\n",
    "    with h5py.File(fname, 'r') as h5file:\n",
    "        beta_orig = np.array(h5file['beta_orig'])\n",
    "        zmat_orig=np.array(h5file['zmat_orig'])\n",
    "        pval_orig = stats.norm.sf(np.abs(zmat_orig))*2.0\n",
    "\n",
    "        betas_snp = beta_orig[:, snps]\n",
    "        pval_snp = pval_orig[:, snps]\n",
    "    \n",
    "        return betas_snp, pval_snp\n",
    "    \n",
    "def data_munger_MOSTEST_betas(snp_df, col_names, cat):\n",
    "    \"\"\"\n",
    "    Load underlying beta's from MOSTEST analysis\n",
    "    \"\"\"\n",
    "    \n",
    "    beta_df = pd.DataFrame(columns=col_names)\n",
    "    pval_df = pd.DataFrame(columns=col_names)\n",
    "    snp_info_df = pd.DataFrame(columns='CHR SNP GP BP A1 A2'.split())\n",
    "    \n",
    "    for snp_no in snp_df.index.values:\n",
    "        snp, snp_all = get_snp(snp_df.chr_no[snp_no], snp_df.ID[snp_no])\n",
    "    \n",
    "        betas_snp, pval_snp = get_betas_pvals(fname=get_file_name_mat(cat, snp_df.chr_no[snp_no]),\n",
    "                                               snps=snp)\n",
    "        \n",
    "        beta_df = add_row(df = beta_df,\n",
    "                          new_data = betas_snp,\n",
    "                          column_names = col_names,\n",
    "                          index_name = snp_df.ID[snp_no])\n",
    "        \n",
    "        pval_df = add_row(df = pval_df,\n",
    "                          new_data = pval_snp,\n",
    "                          column_names = col_names,\n",
    "                          index_name = snp_df.ID[snp_no])\n",
    "        \n",
    "        snp_info_df = add_row(df = snp_info_df,\n",
    "                              new_data = snp_all,\n",
    "                              column_names = 'CHR SNP GP BP A1 A2'.split(),\n",
    "                              index_name = snp_df.ID[snp_no],\n",
    "                              np_array=False)\n",
    "    \n",
    "    return beta_df, pval_df, snp_info_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "699d03b7-919c-4c1d-9912-1aaff39ccbb5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def read_with_gzip(fn):\n",
    "    return pd.read_csv(fn, sep=\"\\t\", compression=\"gzip\", engine=\"pyarrow\")\n",
    "\n",
    "def get_cumulative_pos(df):\n",
    "    running_pos = float(0)\n",
    "\n",
    "    cumulative_pos = []\n",
    "\n",
    "    for chrom, group_df in df.groupby('CHR'):\n",
    "        chr_df = np.add(group_df['BP'].astype(\"int64\"), running_pos)\n",
    "        cumulative_pos.append(chr_df)\n",
    "        running_pos += float(group_df['BP'].max())\n",
    "\n",
    "    df['cumulative_pos'] = pd.concat(cumulative_pos)\n",
    "    \n",
    "    return df\n",
    "\n",
    "def load_MOSTEST_sumstats(fn, out_fn=None):\n",
    "    \n",
    "    data = pd.read_csv(fn, sep=\"\\t\", compression=\"gzip\", engine=\"pyarrow\")\n",
    "    \n",
    "    data = data[data[\"CHR\"]!=\"CHR\"]\n",
    "    \n",
    "    data.loc[data[\"CHR\"] == \"X\", \"CHR\" ] = 23\n",
    "    data.loc[data[\"CHR\"] == \"XY\", \"CHR\" ] = 23\n",
    "    \n",
    "    data[\"CHR\"] = data[\"CHR\"].astype(\"int32\")\n",
    "    data[\"BP\"] = data[\"BP\"].astype(\"int64\")\n",
    "    data[\"PVAL\"] = data[\"PVAL\"].astype(\"float64\")\n",
    "    data[\"Z\"] = data[\"Z\"].astype(\"float32\")\n",
    "    data[\"N\"] = data[\"N\"].astype(\"float32\")\n",
    "    \n",
    "    data.sort_values(by=[\"CHR\", \"BP\"], inplace=True)\n",
    "    \n",
    "    data.reset_index(drop=True, inplace=True)\n",
    "    \n",
    "    #get cumulative position\n",
    "    data = get_cumulative_pos(data)\n",
    "    \n",
    "    data[\"LOGP\"] = -np.log10(data[\"PVAL\"].astype(\"float64\"))\n",
    "\n",
    "    if out_fn is not None:\n",
    "        data.to_csv(out_fn, sep=\"\\t\", compression=\"gzip\")\n",
    "    \n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9c45b818-877f-46fd-a460-09b5262213a7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def MAGMA_data_loader_bs(fn):\n",
    "    return pd.read_csv(fn, delim_whitespace=True, skiprows=5, header=0)\n",
    "\n",
    "def process_bs(fn):\n",
    "    data = MAGMA_data_loader_bs(fn)\n",
    "    data[\"LOGP\"] = -np.log10(data[\"P\"])\n",
    "    return data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eee6abd3-03c1-444a-8ecc-e2314586d896",
   "metadata": {},
   "source": [
    "## Plotting functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c6657395-433f-4daa-a835-ea57887bf9b7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def load_column_names_first_line(fn):\n",
    "    return [str(x) for x in open( fn ).read().split('\\n')[0].split(\"\\t\") ]\n",
    "\n",
    "def load_column_names(fn):\n",
    "    return [str(x) for x in open( fn ).read().split('\\n')[:-1] ] \n",
    "\n",
    "def get_edge_names(list_names):\n",
    "    \"\"\"\n",
    "    Returns edge combinations for edges\n",
    "    \"\"\"\n",
    "    df_names=pd.DataFrame(index=list_names, columns=list_names)\n",
    "    for i in range(len(list_names)):\n",
    "        for j in range(len(list_names)):\n",
    "            df_names.iloc[i, j] = list_names[i]+\"*\"+list_names[j]\n",
    "            \n",
    "    #heritable = load_column_names(\"/data/clusterfs/lag/users/jitame/SENT_CORE/heritable_edges.txt\") + load_column_names(\"/data/clusterfs/lag/users/jitame/SENT_CORE/heritable_edges_asym.txt\")\n",
    "    #heritable = [\"sent_edges_\"+x for x in heritable]\n",
    "    \n",
    "    return [x for x in list(df_names.to_numpy()[np.triu_indices(len(list_names), k=1)].flatten()) if x in heritable]\n",
    "\n",
    "\n",
    "def get_idx(cat):\n",
    "    \"\"\"\n",
    "    Returns indices for category\n",
    "    \"\"\"\n",
    "    sent_core_l_ind = [2, 30, 32, 40, 56, 98, 102, 146, 148, 166, 168, 170, 172, 174, 182, 184, 222, 224]\n",
    "    sent_core_r_ind = [3, 31, 33, 41, 57, 99, 103, 147, 149, 167, 169, 171, 173, 175, 183, 185, 223, 225]\n",
    "    sent_core_bi_ind = sorted(sent_core_l_ind + sent_core_r_ind)\n",
    "    \n",
    "    if cat == \"edges\":\n",
    "        return sent_core_bi_ind\n",
    "    elif cat == \"edges_HD\":\n",
    "        return sent_core_l_ind\n",
    "    \n",
    "    \n",
    "def set_up_coordinates(cat):\n",
    "    \"\"\"\n",
    "    get coordinates for plot\n",
    "    \"\"\"\n",
    "    #set up atlas + coordinates\n",
    "    aicha_atlas = \"/data/workspaces/lag/workspaces/lg-ukbiobank/projects/multilateral/FuncNet_AICHA/segs/AICHA.nii\"\n",
    "    aicha_img = image.load_img(aicha_atlas)\n",
    "    aicha_coords = plotting.find_parcellation_cut_coords(aicha_img)\n",
    "    \n",
    "    #indices\n",
    "    idx = get_idx(cat)\n",
    "\n",
    "    #get coordinates and names\n",
    "    return aicha_coords[idx, :]\n",
    "\n",
    "def get_names_aicha(cat):\n",
    "    \"\"\"\n",
    "    Return node names\n",
    "    \"\"\"\n",
    "    aicha_test = \"/data/clusterfs/lag/projects/lg-ukbiobank/working_data/imaging_data/AICHA/1000099/AICHA_timeseries_NO_GSR_cormat.txt\"\n",
    "    aicha2 = pd.read_csv(aicha_test, sep=\";\", index_col=0)\n",
    "    \n",
    "    idx = get_idx(cat)\n",
    "    \n",
    "    if cat == \"edges\":\n",
    "        return list(aicha2.columns[idx]) #[x for x in list(aicha2.columns[idx]) if x in heritable]\n",
    "    elif cat == \"edges_HD\":\n",
    "        return list(x[:-2] for x in aicha2.columns[idx]) #[x for x in list(aicha2.columns[idx]) if x in heritable]\n",
    "\n",
    "\n",
    "def make_nice_mat(input_data, cat, var_name):\n",
    "    \"\"\"\n",
    "    put data from column into right spot in matrix\n",
    "    \"\"\"\n",
    "    #get indices\n",
    "    ind_names = get_names_aicha(cat)\n",
    "        \n",
    "    #set up dataframe\n",
    "    df_out = pd.DataFrame(index=ind_names, columns=ind_names)\n",
    "    df_out.iloc[np.diag_indices(len(ind_names)), np.diag_indices(len(ind_names))] = 0\n",
    "    \n",
    "    #put data in right place\n",
    "    for x in list(input_data.columns):\n",
    "        nodes = x.split(\"*\", 1)\n",
    "        df_out.loc[nodes[0], nodes[1]] = input_data.loc[var_name, x]\n",
    "        df_out.loc[nodes[1], nodes[0]] = input_data.loc[var_name, x]\n",
    "    \n",
    "    return df_out\n",
    "        \n",
    "    \n",
    "def plot_results_brain_betas(betas, pvals, cat, snp, title, vmax_in=None, alpha=None, ax=None, out=None, rare=False):\n",
    "    \"\"\"\n",
    "    Main function:\n",
    "    - loads data\n",
    "    - significance testing\n",
    "    - plot using nilearn\n",
    "    \"\"\"\n",
    "    #get correlation values\n",
    "    beta_mat = make_nice_mat(betas, cat, snp)\n",
    "    \n",
    "    if alpha:\n",
    "        p_mat = make_nice_mat(pvals, cat, snp)\n",
    "        beta_mat = beta_mat[p_mat < alpha]\n",
    "        \n",
    "    if vmax_in is None:\n",
    "        vmax_in = np.max(np.max(beta_mat))\n",
    "    \n",
    "    #plot\n",
    "    if np.abs(beta_mat.sum().sum()) < 0.0001:\n",
    "        plotting.plot_markers([1]*len(beta_mat),\n",
    "                              set_up_coordinates(cat), \n",
    "                              node_cmap=\"binary\",\n",
    "                              node_size=30,\n",
    "                              alpha=0.8, \n",
    "                              display_mode='lyrz',\n",
    "                              title=title, \n",
    "                              colorbar=False,\n",
    "                              axes=ax)\n",
    "        \n",
    "    else:\n",
    "        if rare:\n",
    "            colors = [\"whitesmoke\", \"indianred\"]\n",
    "            cmap1 = LinearSegmentedColormap.from_list(\"mycmap\", colors)\n",
    "            \n",
    "            plotting.plot_connectome(beta_mat.to_numpy(dtype=float),\n",
    "                             set_up_coordinates(cat),\n",
    "                             title=title,\n",
    "                             edge_cmap=cmap1,\n",
    "                             edge_vmin=0,\n",
    "                             edge_vmax=vmax_in,\n",
    "                             node_color=\"dimgrey\",\n",
    "                             node_size=30,\n",
    "                             display_mode=\"lyrz\",\n",
    "                             colorbar=True,\n",
    "                             alpha=0.8,\n",
    "                             axes=ax)\n",
    "                             #output_file=out)    \n",
    "        else:\n",
    "            colors = [\"mediumblue\", \"cornflowerblue\", \"lightgrey\", \"whitesmoke\", \"lightgrey\", \"lightcoral\", \"indianred\"]\n",
    "            cmap1 = LinearSegmentedColormap.from_list(\"mycmap\", colors)\n",
    "            plotting.plot_connectome(beta_mat.to_numpy(dtype=float),\n",
    "                             set_up_coordinates(cat),\n",
    "                             title=title,\n",
    "                             edge_vmin=-vmax_in,\n",
    "                             edge_vmax=vmax_in,\n",
    "                             edge_cmap=cmap1,\n",
    "                             node_color=\"dimgrey\",\n",
    "                             node_size=30,\n",
    "                             display_mode=\"lyrz\",\n",
    "                             colorbar=True,\n",
    "                             alpha=0.8,\n",
    "                             axes=ax)\n",
    "                             #output_file=out)    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cc4eb207-be66-46c9-a6cc-b71cf811ce02",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def manhattan(df, ax, figargs={\"figsize\":(14,8),\"dpi\":300}):\n",
    "    \n",
    "    g = sns.scatterplot(\n",
    "        data = df,\n",
    "        x = 'cumulative_pos',\n",
    "        y = 'LOGP',\n",
    "        hue = 'CHR',\n",
    "        palette = ['indianred', 'mediumblue'] * 11 + ['indianred'],\n",
    "        linewidth=0,\n",
    "        s=6,\n",
    "        zorder=2,\n",
    "        ax=ax,\n",
    "        legend=None,\n",
    "        style=None\n",
    "    )\n",
    "    \n",
    "    alpha = -np.log10(0.05/10000000)\n",
    "    \n",
    "    ax.plot([np.min(df[\"cumulative_pos\"]), np.max(df[\"cumulative_pos\"])], [alpha, alpha], \"--\", color=\"navy\")\n",
    "\n",
    "    ax.set_xlabel('Chromosome')\n",
    "    ax.set_ylabel('-Log10 P value')\n",
    "\n",
    "    ax.set_xticks(df.groupby('CHR')['cumulative_pos'].median())\n",
    "    ax.tick_params(which='major', width=0.3, length=2, labelsize=8)\n",
    "\n",
    "    ax.set_xticklabels(df['CHR'].unique())\n",
    "    \n",
    "    ax.set_ylim(0, 50)\n",
    "    \n",
    "    ax.spines[\"top\"].set_visible(False)\n",
    "    ax.spines[\"right\"].set_visible(False)\n",
    "    ax.spines[\"left\"].set_visible(True)\n",
    "    ax.spines[\"bottom\"].set_visible(True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5ef0eab3-a20d-46a8-aea0-3cced46ae586",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def gradientbars(bars, ydata, cmap):\n",
    "    \"\"\"\n",
    "    https://towardsdatascience.com/beautiful-bars-scaled-gradient-fill-on-bar-plots-e4da4cdae033\n",
    "    \"\"\"\n",
    "    ax = bars[0].axes\n",
    "    lim = ax.get_xlim()+ax.get_ylim()\n",
    "    ax.axis(lim)\n",
    "    for bar in bars:\n",
    "        bar.set_facecolor(\"none\")\n",
    "        x,y = bar.get_xy()\n",
    "        w, h = bar.get_width(), bar.get_height()\n",
    "        grad = np.atleast_2d(np.linspace(0,1*h/max(ydata),256)).T\n",
    "        ax.imshow(grad, extent=[x,x+w,y,y+h], origin='lower', aspect=\"auto\", \n",
    "                  norm=cm.colors.NoNorm(vmin=0,vmax=1), cmap=plt.get_cmap(cmap))\n",
    "\n",
    "def brainspan_plot(brainspan_data, ax):\n",
    "    colors = [\"lightgray\", \"indianred\"]\n",
    "    cmap1 = LinearSegmentedColormap.from_list(\"mycmap\", colors)\n",
    "    mybar1 = ax.bar(brainspan_data.index.values, brainspan_data[\"LOGP\"], width=0.6)\n",
    "    ax.set_xticks(brainspan_data.index.values, brainspan_data.VARIABLE, rotation=90)\n",
    "    ax.set_xlabel(\"Brainspan age\")\n",
    "    ax.set_ylabel(\"-Log10 P-value of association\")\n",
    "    ax.axhline(-np.log10(0.05/(len(brainspan_data))),color='navy',ls='--') \n",
    "    ax.set_ylim(0, 3.5)\n",
    "    ax.spines[\"top\"].set_visible(False)\n",
    "    ax.spines[\"right\"].set_visible(False)\n",
    "    ax.spines[\"left\"].set_visible(True)\n",
    "    ax.spines[\"bottom\"].set_visible(True)\n",
    "    gradientbars(mybar1, brainspan_data[\"LOGP\"], cmap1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fa84264-f14c-40d9-966d-d69a2933bd02",
   "metadata": {},
   "source": [
    "## Make figure 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cae2106-4f54-4f9b-a65f-734757e32b9a",
   "metadata": {},
   "source": [
    "#### Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fc0b73a9-c8a0-4a9a-895d-75781a0ae59b",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path = \"/data/clusterfs/lag/users/jitame/SENT_CORE/\"\n",
    "results_path = \"/data/workspaces/lag/workspaces/lg-ukbiobank/projects/rest-multimodal/results/\"\n",
    "\n",
    "#mostest mvgwas data\n",
    "mvgwas_lang = load_MOSTEST_sumstats(os.path.join(results_path, \"sumstats\", \"edges.sumstats.gz\"))\n",
    "mvgwas_asym = load_MOSTEST_sumstats(os.path.join(results_path, \"sumstats\", \"edges_asym.sumstats.gz\"))\n",
    "\n",
    "#load data\n",
    "brainspan_lang = process_bs(os.path.join(results_path, \"FUMA\", \"edges_FUMA\", \"magma_exp_bs_age_avg_log2RPKM.gsa.out\"))\n",
    "brainspan_asym = process_bs(os.path.join(results_path, \"FUMA\", \"edges_asym_FUMA\", \"magma_exp_bs_age_avg_log2RPKM.gsa.out\"))\n",
    "\n",
    "#top variants\n",
    "#lang\n",
    "common_top_lang = pd.DataFrame.from_dict({\"chr_no\":[3,3,10], \"ID\":[\"rs35124509\", \"rs2279829\", \"rs2274224\"]})\n",
    "col_names_lang = load_column_names_first_line(\"/data/clusterfs/lag/users/jitame/SENT_CORE/pheno/sent_edges_N29681_resid_mostest.txt\")\n",
    "beta_lang, p_lang, snp_lang = data_munger_MOSTEST_betas(common_top_lang, col_names=col_names_lang, cat=\"edges\")\n",
    "\n",
    "#asym\n",
    "common_top_asym = pd.DataFrame.from_dict({\"chr_no\":[3,3,3], \"ID\":[\"rs7635916\", \"rs2279829\", \"rs13321297\"]})\n",
    "col_names_asym = load_column_names_first_line(\"/data/clusterfs/lag/users/jitame/SENT_CORE/pheno/sent_edges_asym_N29681_resid_mostest.txt\")\n",
    "beta_asym, p_asym, snp_asym = data_munger_MOSTEST_betas(common_top_asym, col_names=col_names_asym, cat=\"edges_HD\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66ade932-c0de-4623-ba40-e1eac9077e06",
   "metadata": {},
   "source": [
    "#### Plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7062b546-d46b-4fe6-88b2-1a40419da2b3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#plot_path = \"/data/workspaces/lag/workspaces/lg-ukbiobank/projects/rest-multimodal/results\"\n",
    "\n",
    "fig_args = {\"figsize\":(17.8,22),\"dpi\":400}\n",
    "\n",
    "fig, ax = plt.subplots(5, 2, gridspec_kw={'height_ratios': [1.4, 1.2, 1, 1, 1]}, **fig_args)\n",
    "\n",
    "vmax_in = float(0.088)\n",
    "vmax_in_hd = float(0.043)\n",
    "\n",
    "#A\n",
    "manhattan(mvgwas_lang, ax=ax[0,0])\n",
    "\n",
    "#B\n",
    "brainspan_plot(brainspan_lang, ax=ax[1,0])\n",
    "\n",
    "#C\n",
    "for x, snp in enumerate(beta_lang.index.values):\n",
    "    plot_results_brain_betas(betas=beta_lang,\n",
    "                             pvals=p_lang,\n",
    "                             vmax_in = vmax_in,\n",
    "                             cat=\"edges\",\n",
    "                             ax=ax[x+2,0],\n",
    "                             title=\"{0} (chr{1}_{2}_{3}:{4})\".format(snp,\n",
    "                                                                   snp_lang.loc[snp_lang.SNP.values == snp, \"CHR\"].to_string(index=False),\n",
    "                                                                   snp_lang.loc[snp_lang.SNP.values == snp, \"BP\"].to_string(index=False),\n",
    "                                                                   snp_lang.loc[snp_lang.SNP.values == snp, \"A1\"].to_string(index=False),\n",
    "                                                                   snp_lang.loc[snp_lang.SNP.values == snp, \"A2\"].to_string(index=False)),\n",
    "                             snp=snp)\n",
    "\n",
    "#D\n",
    "manhattan(mvgwas_asym, ax=ax[0,1])\n",
    "\n",
    "#E\n",
    "brainspan_plot(brainspan_asym, ax=ax[1,1])\n",
    "\n",
    "#F\n",
    "for x, snp in enumerate(beta_asym.index.values):\n",
    "    plot_results_brain_betas(betas=beta_asym,\n",
    "                             pvals=p_asym,\n",
    "                             cat=\"edges_HD\",\n",
    "                             vmax_in = vmax_in_hd,\n",
    "                             ax=ax[x+2,1],\n",
    "                             snp=snp,\n",
    "                             title=\"{0} (chr{1}_{2}_{3}:{4})\".format(snp,\n",
    "                                                                     snp_asym.loc[snp_asym.SNP.values == snp, \"CHR\"].to_string(index=False),\n",
    "                                                                     snp_asym.loc[snp_asym.SNP.values == snp, \"BP\"].to_string(index=False),\n",
    "                                                                     snp_asym.loc[snp_asym.SNP.values == snp, \"A1\"].to_string(index=False),\n",
    "                                                                     snp_asym.loc[snp_asym.SNP.values == snp, \"A2\"].to_string(index=False)))\n",
    "\n",
    "\n",
    "plt.subplots_adjust(hspace=0.5, wspace=0.1)\n",
    "\n",
    "gwas_ticks=[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,\"\",17,\"\",19,\"\",21,\"\",\"X\"]\n",
    "\n",
    "#Titles\n",
    "#fig.suptitle(\"Common variant results\", y=0.95, fontsize=28)\n",
    "ax[0, 0].text(50, 61, \"Language network\", fontsize=20)\n",
    "ax[0, 0].set_xticks(ax[0,0].get_xticks(), gwas_ticks)\n",
    "ax[0, 0].text(0, 55, \"A. mvGWAS\", fontsize=16)\n",
    "ax[1, 0].text(0, 3.85, \"C. Brainspan MAGMA\", fontsize=16)\n",
    "ax[2, 0].text(0.06, 1.08, \"E. Betas lead variants\", fontsize=16)\n",
    "ax[0, 1].text(50, 61, \"Hemispheric differences\", fontsize=20)\n",
    "ax[0, 1].set_xticks(ax[0,1].get_xticks(), gwas_ticks)\n",
    "ax[0, 1].text(0, 55, \"B. mvGWAS\", fontsize=16)\n",
    "ax[1, 1].text(0, 3.85, \"D. Brainspan MAGMA\", fontsize=16)\n",
    "ax[2, 1].text(0.06, 1.08, \"F. Betas lead variants\", fontsize=16)\n",
    "\n",
    "\n",
    "#plt.savefig(fname=os.path.join(results_path, \"Figure_1_common_results_v3.tif\"), bbox_inches=\"tight\")\n",
    "#plt.savefig(fname=os.path.join(results_path, \"Figure_1_common_results_v3.eps\"), bbox_inches=\"tight\")\n",
    "plt.savefig(fname=os.path.join(results_path, \"Figure_1_common_results_v3.pdf\"), bbox_inches=\"tight\")\n",
    "plt.savefig(fname=os.path.join(results_path, \"Figure_1_common_results_v3.png\"), bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f225585d-feb7-4040-817a-48064efc79d8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:results_env]",
   "language": "python",
   "name": "conda-env-results_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
