{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f7fc83d9-57ff-47b3-9f1a-d7ef5d40b61f",
   "metadata": {},
   "source": [
    "# Figure 3. Rare variant results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad1f43ab-f32a-43ed-8342-ac6035156f10",
   "metadata": {},
   "source": [
    "### Dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38af86d8-73e8-42e7-bc70-5b90b641e6f4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import re\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.io as sio\n",
    "import scipy.stats as stats\n",
    "from statsmodels.nonparametric.kernel_regression import KernelReg\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from nilearn import plotting, image\n",
    "from matplotlib import cm\n",
    "from matplotlib.colors import LinearSegmentedColormap"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72998f0a-dc5b-44b5-8893-01a0ad1bce97",
   "metadata": {},
   "source": [
    "### Plotting "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e304247-1d93-495a-bd2e-96ad207e1119",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def load_column_names_first_line(fn):\n",
    "    return [str(x) for x in open( fn ).read().split('\\n')[0].split(\"\\t\") ]\n",
    "\n",
    "def load_column_names(fn):\n",
    "    return [str(x) for x in open( fn ).read().split('\\n')[:-1] ] \n",
    "\n",
    "def get_edge_names(list_names):\n",
    "    \"\"\"\n",
    "    Returns edge combinations for edges\n",
    "    \"\"\"\n",
    "    df_names=pd.DataFrame(index=list_names, columns=list_names)\n",
    "    for i in range(len(list_names)):\n",
    "        for j in range(len(list_names)):\n",
    "            df_names.iloc[i, j] = list_names[i]+\"*\"+list_names[j]\n",
    "            \n",
    "    #heritable = load_column_names(\"/data/clusterfs/lag/users/jitame/SENT_CORE/heritable_edges.txt\") + load_column_names(\"/data/clusterfs/lag/users/jitame/SENT_CORE/heritable_edges_asym.txt\")\n",
    "    #heritable = [\"sent_edges_\"+x for x in heritable]\n",
    "    \n",
    "    return [x for x in list(df_names.to_numpy()[np.triu_indices(len(list_names), k=1)].flatten()) if x in heritable]\n",
    "\n",
    "\n",
    "def get_idx(cat):\n",
    "    \"\"\"\n",
    "    Returns indices for category\n",
    "    \"\"\"\n",
    "    sent_core_l_ind = [2, 30, 32, 40, 56, 98, 102, 146, 148, 166, 168, 170, 172, 174, 182, 184, 222, 224]\n",
    "    sent_core_r_ind = [3, 31, 33, 41, 57, 99, 103, 147, 149, 167, 169, 171, 173, 175, 183, 185, 223, 225]\n",
    "    sent_core_bi_ind = sorted(sent_core_l_ind + sent_core_r_ind)\n",
    "    \n",
    "    if cat == \"edges\":\n",
    "        return sent_core_bi_ind\n",
    "    elif cat == \"edges_HD\":\n",
    "        return sent_core_l_ind\n",
    "    \n",
    "    \n",
    "def set_up_coordinates(cat):\n",
    "    \"\"\"\n",
    "    get coordinates for plot\n",
    "    \"\"\"\n",
    "    #set up atlas + coordinates\n",
    "    aicha_atlas = \"/data/workspaces/lag/workspaces/lg-ukbiobank/projects/multilateral/FuncNet_AICHA/segs/AICHA.nii\"\n",
    "    aicha_img = image.load_img(aicha_atlas)\n",
    "    aicha_coords = plotting.find_parcellation_cut_coords(aicha_img)\n",
    "    \n",
    "    #indices\n",
    "    idx = get_idx(cat)\n",
    "\n",
    "    #get coordinates and names\n",
    "    return aicha_coords[idx, :]\n",
    "\n",
    "def get_names_aicha(cat):\n",
    "    \"\"\"\n",
    "    Return node names\n",
    "    \"\"\"\n",
    "    aicha_test = \"/data/clusterfs/lag/projects/lg-ukbiobank/working_data/imaging_data/AICHA/1000099/AICHA_timeseries_NO_GSR_cormat.txt\"\n",
    "    aicha2 = pd.read_csv(aicha_test, sep=\";\", index_col=0)\n",
    "    \n",
    "    idx = get_idx(cat)\n",
    "    \n",
    "    if cat == \"edges\":\n",
    "        return list(aicha2.columns[idx]) #[x for x in list(aicha2.columns[idx]) if x in heritable]\n",
    "    elif cat == \"edges_HD\":\n",
    "        return list(x[:-2] for x in aicha2.columns[idx]) #[x for x in list(aicha2.columns[idx]) if x in heritable]\n",
    "\n",
    "\n",
    "def make_nice_mat(input_data, cat, var_name):\n",
    "    \"\"\"\n",
    "    put data from column into right spot in matrix\n",
    "    \"\"\"\n",
    "    #get indices\n",
    "    ind_names = get_names_aicha(cat)\n",
    "        \n",
    "    #set up dataframe\n",
    "    df_out = pd.DataFrame(index=ind_names, columns=ind_names)\n",
    "    df_out.iloc[np.diag_indices(len(ind_names)), np.diag_indices(len(ind_names))] = 0\n",
    "    \n",
    "    #put data in right place\n",
    "    for x in list(input_data.columns):\n",
    "        nodes = x.split(\"*\", 1)\n",
    "        df_out.loc[nodes[0], nodes[1]] = input_data.loc[var_name, x]\n",
    "        df_out.loc[nodes[1], nodes[0]] = input_data.loc[var_name, x]\n",
    "    \n",
    "    return df_out\n",
    "        \n",
    "    \n",
    "def plot_results_brain_betas(betas, pvals, cat, snp, title, alpha=None, ax=None, out=None, rare=False):\n",
    "    \"\"\"\n",
    "    Main function:\n",
    "    - loads data\n",
    "    - significance testing\n",
    "    - plot using nilearn\n",
    "    \"\"\"\n",
    "    #get correlation values\n",
    "    beta_mat = make_nice_mat(betas, cat, snp)\n",
    "    \n",
    "    if alpha:\n",
    "        p_mat = make_nice_mat(pvals, cat, snp)\n",
    "        beta_mat = beta_mat[p_mat < alpha]\n",
    "    \n",
    "    #plot\n",
    "    if np.abs(beta_mat.sum().sum()) < 0.0001:\n",
    "        plotting.plot_markers([1]*len(beta_mat),\n",
    "                              set_up_coordinates(cat), \n",
    "                              node_cmap=\"binary\",\n",
    "                              node_size=30,\n",
    "                              alpha=0.8, \n",
    "                              display_mode='lyrz',\n",
    "                              title=title, \n",
    "                              colorbar=False,\n",
    "                              axes=ax)\n",
    "        \n",
    "    else:\n",
    "        if rare:\n",
    "            colors = [\"whitesmoke\", \"indianred\"]\n",
    "            cmap1 = LinearSegmentedColormap.from_list(\"mycmap\", colors)\n",
    "            \n",
    "            plotting.plot_connectome(beta_mat.to_numpy(dtype=float),\n",
    "                             set_up_coordinates(cat),\n",
    "                             title=title,\n",
    "                             edge_cmap=cmap1,\n",
    "                             edge_vmin=0,\n",
    "                             edge_vmax=np.max(np.max(beta_mat)),\n",
    "                             node_color=\"dimgrey\",\n",
    "                             node_size=30,\n",
    "                             display_mode=\"lyrz\",\n",
    "                             colorbar=True,\n",
    "                             alpha=0.8,\n",
    "                             axes=ax)\n",
    "                             #output_file=out)    \n",
    "        else:\n",
    "            colors = [\"mediumblue\", \"cornflowerblue\", \"lightgrey\", \"whitesmoke\", \"lightgrey\", \"lightcoral\", \"indianred\"]\n",
    "            cmap1 = LinearSegmentedColormap.from_list(\"mycmap\", colors)\n",
    "            plotting.plot_connectome(beta_mat.to_numpy(dtype=float),\n",
    "                             set_up_coordinates(cat),\n",
    "                             title=title,\n",
    "                             edge_cmap=cmap1,\n",
    "                             node_color=\"dimgrey\",\n",
    "                             node_size=30,\n",
    "                             display_mode=\"lyrz\",\n",
    "                             colorbar=True,\n",
    "                             alpha=0.8,\n",
    "                             axes=ax)\n",
    "                             #output_file=out)    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b894a45a-2b96-4f47-8613-1c4e854436a2",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def plot_brainspan(data, genes, unique_columns, colors, ax, bw):\n",
    "    \"\"\"\n",
    "    Plots brainspan data with columns and genes of interest\n",
    "    \n",
    "    Dependencies:\n",
    "    matplotlib.pyplot\n",
    "    numpy\n",
    "    KernelReg from statsmodels (for smoother curve)\n",
    "    \"\"\"\n",
    "    for y_gene, gene in enumerate(genes):\n",
    "        avgs = []\n",
    "        stds = []\n",
    "        color = colors[y_gene]\n",
    "\n",
    "        for x, age in enumerate(unique_columns):\n",
    "            y = data.loc[gene, age].to_numpy().flatten()\n",
    "            ax.scatter(x=[x]*len(y), y=y, color=color, s=.6)\n",
    "            avgs.append(np.mean(y))\n",
    "            stds.append(np.std(y))\n",
    "\n",
    "        #do a bit of smoothing of the averages using kernel regression\n",
    "        x = np.arange(len(unique_columns))\n",
    "        kr = KernelReg(np.array(avgs), x, 'c', bw=bw)\n",
    "        avgs_pred, _ = kr.fit(x)\n",
    "        kr_std = KernelReg(np.array(stds), x, 'c', bw=bw)\n",
    "        std_pred, _ = kr_std.fit(x)\n",
    "\n",
    "        low_err = avgs_pred-std_pred.flatten()\n",
    "        high_err = avgs_pred + std_pred.flatten()\n",
    "\n",
    "        ax.plot(x, avgs_pred, color=color, label=gene)\n",
    "        ax.fill_between(x, low_err, high_err, alpha=0.2, edgecolor=color, facecolor=color)\n",
    "\n",
    "        #TODO\n",
    "        ax.set_xlabel(\"Developmental age\")\n",
    "        ax.set_ylabel(\"RNA-Seq RPKM values\")\n",
    "        \n",
    "        ax.set_xticks(np.arange(len(unique_columns)), unique_columns, rotation=60)\n",
    "        ax.legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a4d2e87-f22c-4b1d-8af2-3d9edd3fc52c",
   "metadata": {},
   "source": [
    "### Data loaders: rare variants + Brainspan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90b97728-cdf1-4485-b70a-7a57b52203fd",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def add_gene_names(data):\n",
    "    \"\"\"\n",
    "    ADDS GENE NAME BASED ON ENSEMBL ID\n",
    "    \"\"\"\n",
    "    \n",
    "    from pyensembl import EnsemblRelease\n",
    "    data_ens = EnsemblRelease(108)\n",
    "    \n",
    "    #parse gene ids and names\n",
    "    data[\"gene_id\"] = [id_name.split(\".\")[0] for id_name in data[\"ID\"]]\n",
    "    data[\"gene_name\"] = [get_gene_name(gene_id, data_ens) for gene_id in data[\"gene_id\"]]\n",
    "    \n",
    "    return data\n",
    "\n",
    "def get_gene_name(gene_id, data_ens):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    try:\n",
    "        gene_name = data_ens.gene_name_of_gene_id(gene_id)\n",
    "    except ValueError:\n",
    "        gene_name = \"\"\n",
    "    return gene_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07c04d4b-baa4-4971-b47c-1641490db5d9",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def data_loader_brainspan(in_path, genes_in):\n",
    "    #load columns and rows\n",
    "    columns = pd.read_csv(os.path.join(in_path, \"columns_metadata.csv\"))\n",
    "    rows = pd.read_csv(os.path.join(in_path, \"rows_metadata.csv\"))\n",
    "    \n",
    "    #check if in Brainspan data\n",
    "    genes_to_keep = [x for x in genes_in if x in rows.gene_symbol.unique()]\n",
    "\n",
    "    #Boolean for genes_in\n",
    "    boolean_keep = [x in genes_to_keep for x in rows.gene_symbol]\n",
    "    \n",
    "    #get rows\n",
    "    rows_with_genes = rows.loc[boolean_keep, :]\n",
    "    row_nos = rows_with_genes.index.values\n",
    "\n",
    "    #make multiple index\n",
    "    arrays = [columns.age.to_numpy(), columns.structure_acronym.to_numpy()]\n",
    "    \n",
    "    #only read relevant rows\n",
    "    data = pd.read_csv(os.path.join(in_path, \"expression_matrix.csv\"), skiprows = lambda x: x not in row_nos, header=None, index_col = 0)\n",
    "    data = pd.DataFrame(data.to_numpy(), columns=arrays, index=rows_with_genes.gene_symbol)\n",
    "    data.columns.names=[\"Age\", \"Structure\"]\n",
    "    \n",
    "    age_columns = columns.age.unique()\n",
    "    \n",
    "    return data, age_columns, list(rows_with_genes.gene_symbol)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a20d3ff-19b8-4218-8e4a-4e10f3bfcacc",
   "metadata": {},
   "source": [
    "### Data munging figure 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b90236f0-77c8-4fbf-8e71-771eb246a381",
   "metadata": {},
   "outputs": [],
   "source": [
    "#set significance testing\n",
    "alpha_burden_genebass = -np.log10(6.7 * 10 ** -7)\n",
    "alpha_skato_genebass = -np.log10(2.5 * 10 ** -7) #from genebass paper: https://www.cell.com/cell-genomics/fulltext/S2666-979X(22)00110-0#secsectitle0080\n",
    "\n",
    "#set figure path\n",
    "plot_path = \"/data/workspaces/lag/workspaces/lg-ukbiobank/projects/rest-multimodal/results\"\n",
    "\n",
    "#data munging\n",
    "rare_genes = pd.read_csv(os.path.join(\"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/regenie_gene_based/\", \"all_exome_results.csv\"), engine=\"pyarrow\")\n",
    "rare_genes = add_gene_names(rare_genes)\n",
    "\n",
    "data_skato_strict = rare_genes[np.array([rare_genes[\"TEST\"] == \"ADD-SKATO\", rare_genes[\"ALLELE1\"] == \"Strict.0.01\"]).all(0).reshape((len(rare_genes), 1))]\n",
    "data_skato_broad = rare_genes[np.array([rare_genes[\"TEST\"] == \"ADD-SKATO\", rare_genes[\"ALLELE1\"] == \"Broad.0.01\"]).all(0).reshape((len(rare_genes), 1))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe0b1def-f8ce-4de3-b2eb-d66b582ff0f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#data mugging 2\n",
    "#get lang net results\n",
    "heritable_lang = load_column_names(\"/data/clusterfs/lag/users/jitame/SENT_CORE/heritable_edges.txt\")\n",
    "heritable_lang = [\"LOGP_sent_edges_\"+x for x in heritable_lang]\n",
    "lang_strict_sig = data_skato_strict[data_skato_strict[\"MINP_edges\"] > alpha_skato_genebass].set_index(\"gene_name\")\n",
    "lang_broad_sig = data_skato_broad[data_skato_broad[\"MINP_edges\"] > alpha_skato_genebass].set_index(\"gene_name\")\n",
    "\n",
    "p_df_broad = lang_broad_sig[heritable_lang]\n",
    "p_df_broad.columns=[x[16:] for x in p_df_broad.columns]\n",
    "\n",
    "#get HD results\n",
    "heritable_asym = load_column_names(\"/data/clusterfs/lag/users/jitame/SENT_CORE/heritable_edges_asym.txt\")\n",
    "heritable_asym = [\"LOGP_sent_edges_\"+x for x in heritable_asym]\n",
    "asym_strict_sig = data_skato_strict[data_skato_strict[\"MINP_HDs\"] > alpha_skato_genebass].set_index(\"gene_name\")\n",
    "asym_broad_sig = data_skato_broad[data_skato_broad[\"MINP_HDs\"] > alpha_skato_genebass].set_index(\"gene_name\")\n",
    "\n",
    "p_df_strict = asym_strict_sig[heritable_asym]\n",
    "p_df_strict.columns=[x[16:] for x in p_df_strict.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6b51ea1-8141-4df7-803c-b2c92aa475f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#data munging 3 \n",
    "in_path = \"/data/workspaces/lag/workspaces/lg-ukbiobank/projects/rest-multimodal/brainspan\"\n",
    "\n",
    "#specify genes of interest\n",
    "genes_in = [\"MANEAL\", \"DUSP29\",  \"SLC25A48\", \"TRIP11\", \"NIBAN1\",\n",
    "                     \"DDX25\", \"WDCP\"]\n",
    "\n",
    "data_bs, age_columns, genes_bs  = data_loader_brainspan(in_path, genes_in)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3cec3bc2-a0e8-426c-a322-b169a7837a2f",
   "metadata": {},
   "source": [
    "### Make the plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88c912b7-7ef4-4d6b-a245-310d1b29edfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_args = {\"figsize\":(18,24),\"dpi\":300}\n",
    "\n",
    "colors_bs = [\"cornflowerblue\", \"indianred\",  \"mediumblue\", \"lightcoral\"]\n",
    "\n",
    "fig, ax = plt.subplots(5, 2, gridspec_kw={'height_ratios': [1]*len(p_df_broad)}, **fig_args)\n",
    "\n",
    "#plot language network results\n",
    "for x, gene in enumerate(p_df_broad.index.values):\n",
    "    plot_results_brain_betas(betas=p_df_broad,\n",
    "                             pvals=p_df_broad,\n",
    "                             cat=\"edges\",\n",
    "                             ax=ax[x, 0],\n",
    "                             title=gene,\n",
    "                             snp=gene,\n",
    "                            rare=True)\n",
    "\n",
    "#plot hemispheric differences\n",
    "for x, gene in enumerate(p_df_strict.index.values):\n",
    "    plot_results_brain_betas(betas=p_df_strict,\n",
    "                             pvals=p_df_strict,\n",
    "                             cat=\"edges_HD\",\n",
    "                             ax=ax[x, 1],\n",
    "                             title=gene,\n",
    "                             snp=gene,\n",
    "                             rare=True)\n",
    "\n",
    "#plot densities\n",
    "sns.kdeplot(p_df_broad.T, ax=ax[2,1])\n",
    "sns.kdeplot(p_df_strict.T, ax=ax[3,1])\n",
    "\n",
    "#plot brainspan expression data\n",
    "plot_brainspan(data_bs,\n",
    "               genes_bs,\n",
    "               age_columns,\n",
    "               colors_bs,\n",
    "               ax[4,1],\n",
    "               bw=[2.5])\n",
    "\n",
    "#adjust spacing\n",
    "plt.subplots_adjust(hspace=0.3, wspace=0.1)\n",
    "\n",
    "#set spines\n",
    "ax[2,1].spines[\"top\"].set_visible(False)\n",
    "ax[2,1].spines[\"right\"].set_visible(False)\n",
    "ax[2,1].spines[\"left\"].set_visible(True)\n",
    "ax[2,1].spines[\"bottom\"].set_visible(True)\n",
    "ax[3,1].spines[\"top\"].set_visible(False)\n",
    "ax[3,1].spines[\"right\"].set_visible(False)\n",
    "ax[3,1].spines[\"left\"].set_visible(True)\n",
    "ax[3,1].spines[\"bottom\"].set_visible(True)\n",
    "ax[4,1].spines[\"top\"].set_visible(False)\n",
    "ax[4,1].spines[\"right\"].set_visible(False)\n",
    "ax[4,1].spines[\"left\"].set_visible(True)\n",
    "ax[4,1].spines[\"bottom\"].set_visible(True)\n",
    "\n",
    "#titles\n",
    "fig.suptitle(\"Rare variant results\", y=0.95, fontsize=28)\n",
    "ax[0, 0].text(0, 1.13, \"A. -Log10 P SKATO language network results with broad filter\", fontsize=16)\n",
    "ax[0, 1].text(0, 1.13, \"B. -Log10 P SKATO hemispheric differences results with strict filter\", fontsize=16)\n",
    "ax[2,1].text(-1, 0.25, \"C. Density plots for gene language network edges associations\", fontsize=16)\n",
    "ax[3,1].text(-1, 0.45, \"D. Density plots for gene hemispheric differences associations\", fontsize=16)\n",
    "ax[4,1].text(-1, 40, \"E. Brainspan gene expression data\", fontsize=16)\n",
    "\n",
    "plt.savefig(fname=os.path.join(plot_path, \"Fig_3_rare_variant_results.png\"), bbox_inches=\"tight\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "380dd125-ef7a-4ffa-b427-c4c3083733c5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:results_env]",
   "language": "python",
   "name": "conda-env-results_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
