{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b45eab04-8eb5-4037-b56a-633ca8cea128",
   "metadata": {},
   "source": [
    "# Reviewer response formal tests global patterns\n",
    "\n",
    "In this notebook we formally test whether the genetic associations are specific or not.\n",
    "\n",
    "Inputs:\n",
    "- z scores from mvGWAS from MOSTEST (because they take SE into account)\n",
    "- CCA loadings from CCA with polygenic scores\n",
    "- z scores (beta/se) from burden regression\n",
    "\n",
    "We use an independent samples t-test to test differences between:\n",
    "- Left intra-hemispheric connectivity vs. right intra-hemispheric connectivity\n",
    "- Intra-hemispheric vs. interhemispheric connectivity\n",
    "\n",
    "We use a one-sample t-test to test whether the overall pattern is significantly different from zero\n",
    "\n",
    "We output to csv-files.\n",
    "\n",
    "Last update: 2024-06-26, JS Amelink"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0c6e5e21-272a-4bb4-9049-11595c344a0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import h5py\n",
    "import re\n",
    "from scipy.stats import ttest_ind, ttest_1samp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5fd5904-b517-4065-9428-46b7e6cb478f",
   "metadata": {},
   "source": [
    "## Helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f43e0528-1c7a-4e78-b3c7-0736f57c0265",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def decide_pheno(brain_pheno):\n",
    "    \"\"\"\n",
    "    Decides which phenotype category based on filename.\n",
    "    Input:\n",
    "    - brain pheno name\n",
    "    Output:\n",
    "    - category\n",
    "    \n",
    "    Categories:\n",
    "    edges, edges HD\n",
    "    \"\"\"\n",
    "        \n",
    "        #split edge name\n",
    "    fn_split=brain_pheno.split('*')\n",
    "        \n",
    "        #conditional for categorization:\n",
    "    if re.search(\"-L\", fn_split[0]) is not None and re.search(\"-L\", fn_split[1]) is not None:\n",
    "        sub = 'L'\n",
    "    elif re.search(\"-R\", fn_split[0]) is not None and re.search(\"-R\", fn_split[1]) is not None:\n",
    "        sub = 'R'\n",
    "    elif re.search(\"-L\", fn_split[0]) is not None and re.search(\"-R\", fn_split[1]) is not None:\n",
    "        sub = 'inter'\n",
    "    elif re.search(\"-R\", fn_split[0]) is not None and re.search(\"-L\", fn_split[1]) is not None:\n",
    "        sub = 'inter'\n",
    "    else:\n",
    "        sub = 'L-R'\n",
    "                \n",
    "    return sub\n",
    "\n",
    "def add_row(df, new_data, column_names, index_name, np_array=True):\n",
    "    if np_array:\n",
    "        df_new = pd.DataFrame(data=new_data.reshape((1, len(column_names))), columns=column_names, index=[index_name])\n",
    "    else:\n",
    "        df_new = pd.DataFrame.from_dict(new_data)\n",
    "    return pd.concat([df, df_new], axis=0)\n",
    "        \n",
    "def get_snp(chr_no, snp_id):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    if chr_no == 23:\n",
    "        bim_fn = \"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/mostest/in/mostest_geno_in_cXY.bim\"\n",
    "    else:\n",
    "        bim_fn = \"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/mostest/in/mostest_geno_in_c{}.bim\".format(chr_no)\n",
    "    \n",
    "    bim = pd.read_csv(bim_fn, sep='\\t', header=None, names='CHR SNP GP BP A1 A2'.split())\n",
    "    snp = bim.SNP.values == snp_id\n",
    "    return snp, bim.loc[bim.SNP.values == snp_id, :].to_dict()\n",
    "\n",
    "def get_file_name_mat(cat, chr_no):\n",
    "    if chr_no == 23:\n",
    "        return \"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/mostest/out/sent_edges/edges_chrXY_zmat.mat\"\n",
    "    if cat == \"edges\":\n",
    "        return \"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/mostest/out/sent_edges/edges_chr{}_zmat.mat\".format(chr_no)\n",
    "    if cat == \"edges_HD\":\n",
    "        return \"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/mostest/out/sent_edges_asym/edges_asym_chr{}_zmat.mat\".format(chr_no)\n",
    "\n",
    "def get_betas_zs(fname, snps):\n",
    "    with h5py.File(fname, 'r') as h5file:\n",
    "        beta_orig = np.array(h5file['beta_orig'])\n",
    "        zmat_orig=np.array(h5file['zmat_orig'])\n",
    "        #pval_orig = stats.norm.sf(np.abs(zmat_orig))*2.0\n",
    "\n",
    "        betas_snp = beta_orig[:, snps]\n",
    "        z_snp = zmat_orig[:, snps]\n",
    "    \n",
    "        return betas_snp, z_snp\n",
    "    \n",
    "def data_munger_MOSTEST_betas(snp_df, col_names, cat):\n",
    "    \"\"\"\n",
    "    Load underlying beta's from MOSTEST analysis\n",
    "    \"\"\"\n",
    "    \n",
    "    beta_df = pd.DataFrame(columns=col_names)\n",
    "    z_df = pd.DataFrame(columns=col_names)\n",
    "    snp_info_df = pd.DataFrame(columns='CHR SNP GP BP A1 A2'.split())\n",
    "    \n",
    "    for snp_no in snp_df.index.values:\n",
    "        snp, snp_all = get_snp(snp_df.chr_no[snp_no], snp_df.ID[snp_no])\n",
    "    \n",
    "        betas_snp, pval_snp = get_betas_zs(fname=get_file_name_mat(cat, snp_df.chr_no[snp_no]),\n",
    "                                               snps=snp)\n",
    "        \n",
    "        beta_df = add_row(df = beta_df,\n",
    "                          new_data = betas_snp,\n",
    "                          column_names = col_names,\n",
    "                          index_name = snp_df.ID[snp_no])\n",
    "        \n",
    "        z_df = add_row(df = z_df,\n",
    "                          new_data = pval_snp,\n",
    "                          column_names = col_names,\n",
    "                          index_name = snp_df.ID[snp_no])\n",
    "        \n",
    "        snp_info_df = add_row(df = snp_info_df,\n",
    "                              new_data = snp_all,\n",
    "                              column_names = 'CHR SNP GP BP A1 A2'.split(),\n",
    "                              index_name = snp_df.ID[snp_no],\n",
    "                              np_array=False)\n",
    "    \n",
    "    return beta_df, z_df, snp_info_df\n",
    "\n",
    "def load_column_names_first_line(fn):\n",
    "    return [str(x) for x in open( fn ).read().split('\\n')[0].split(\"\\t\") ]\n",
    "\n",
    "def ttest_all(df):\n",
    "    genos = list(df.columns)[:-1]\n",
    "    alpha = 0.05 / ((len(df.columns)-1)*3)\n",
    "\n",
    "    a = df[df[\"Category\"] == \"L\"].iloc[:,:-1].to_numpy(dtype=np.float64)\n",
    "    b = df[df[\"Category\"] == \"R\"].iloc[:,:-1].to_numpy(dtype=np.float64)\n",
    "    c = np.vstack([a, b])\n",
    "    d = df[df[\"Category\"] == \"inter\"].iloc[:,:-1].to_numpy(dtype=np.float64)\n",
    "    e = np.vstack([c,d])\n",
    "\n",
    "    mean_l = np.mean(a, axis=0)\n",
    "    mean_r = np.mean(b, axis=0)\n",
    "    mean_intra = np.mean(c, axis=0)\n",
    "    mean_inter = np.mean(d, axis=0)\n",
    "    mean_all = np.mean(e, axis=0)\n",
    "    \n",
    "    std_l = np.std(a, axis=0)\n",
    "    std_r = np.std(b, axis=0)\n",
    "    std_intra = np.std(c, axis=0)\n",
    "    std_inter = np.std(d, axis=0)\n",
    "    std_all = np.std(e, axis=0)\n",
    "    \n",
    "    #test different from 0\n",
    "    global_t, global_p = ttest_1samp(e, popmean=0)\n",
    "    \n",
    "    #test differences\n",
    "    lr_t, lr_p = ttest_ind(a, b)\n",
    "    int_t, int_p = ttest_ind(c, d)\n",
    "    \n",
    "    #bonferroni correction\n",
    "    bonf_global = global_p < alpha\n",
    "    bonf_lr = lr_p < alpha\n",
    "    bonf_int = int_p < alpha\n",
    "    \n",
    "    log10p_global = -np.log10(global_p)\n",
    "    log10p_lr = -np.log10(lr_p)\n",
    "    log10p_int = -np.log10(int_p)\n",
    "    \n",
    "    t_df = pd.DataFrame([mean_all, std_all, global_t, log10p_global, bonf_global, mean_l, std_l, mean_r, std_r, lr_t, log10p_lr, bonf_lr, mean_intra, std_intra, mean_inter, std_inter, int_t, log10p_int, bonf_int]).T\n",
    "    t_df.index = genos\n",
    "    t_df.columns = [\"Mean_all\", \"Std_all\", \"T_test_1samp_all\", \"Log10P_all\", \"Significance_all\", \"Mean_L\", \"Std_L\", \"Mean_R\", \"Std_R\", \"T_test_LR\", \"Log10P_LR\", \"Signficant_LR\", \"Mean_intra\", 'Std_intra', \"Mean_inter\", \"Std_inter\", \"T_test_intra_inter\", \"Log10P_intra_inter\", \"Significant_intra_inter\"]\n",
    "\n",
    "    return t_df\n",
    "\n",
    "def ttest_asym(df):\n",
    "    alpha = 0.05/df.shape[1]\n",
    "    print(\"Alpha: {}\".format(alpha))\n",
    "    genos = list(df.columns)\n",
    "\n",
    "    z = df.to_numpy(dtype=np.float64)\n",
    "\n",
    "    mean_z = np.mean(z, axis=0)\n",
    "    std_z = np.std(z, axis=0)\n",
    "    t, p = ttest_1samp(z, popmean=0)\n",
    "\n",
    "    log10p = -np.log10(p)\n",
    "    \n",
    "    sig = p < alpha\n",
    "    print(p)\n",
    "\n",
    "    t_df = pd.DataFrame([mean_z, std_z, t, log10p, sig]).T\n",
    "    t_df.index = genos\n",
    "    t_df.columns = [\"Mean_all\", \"Std_all\", \"T_test_1samp_all\", \"Log10P_all\", \"Significance\"]\n",
    "    \n",
    "    return t_df \n",
    "\n",
    "def add_gene_names(data):\n",
    "    \"\"\"\n",
    "    ADDS GENE NAME BASED ON ENSEMBL ID\n",
    "    \"\"\"\n",
    "    \n",
    "    from pyensembl import EnsemblRelease\n",
    "    data_ens = EnsemblRelease(108)\n",
    "    \n",
    "    #parse gene ids and names\n",
    "    data[\"gene_id\"] = [id_name.split(\".\")[0] for id_name in data[\"ID\"]]\n",
    "    data[\"gene_name\"] = [get_gene_name(gene_id, data_ens) for gene_id in data[\"gene_id\"]]\n",
    "    \n",
    "    return data\n",
    "\n",
    "def get_gene_name(gene_id, data_ens):\n",
    "    \"\"\"\n",
    "    \"\"\"\n",
    "    try:\n",
    "        gene_name = data_ens.gene_name_of_gene_id(gene_id)\n",
    "    except ValueError:\n",
    "        gene_name = \"\"\n",
    "    return gene_name\n",
    "\n",
    "def load_column_names(fn):\n",
    "    return [str(x) for x in open( fn ).read().split('\\n')[:-1]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "637142f5-92d0-4e37-b780-f3059ad7e810",
   "metadata": {},
   "source": [
    "## Language network: SNPS + polygenic scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "843a5f21-5d43-4ca9-8cde-936c64ba3f2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#load snp data\n",
    "common_all_lang = pd.DataFrame.from_dict({\"chr_no\":[2,3,3,3,3,4,10,10,11,14,17,17,18,23], \"ID\":[\"rs2717066\", \"rs17273020\", \"rs35124509\", \"rs143322006\", \"rs2279829\", \"rs2131466\", \"rs2274224\", \"rs4469802\", \"rs10897597\", \"rs468213\", \"rs4924967\", \"rs2732687\", \"rs7234875\", \"rs2360257\"]})\n",
    "col_names_lang = load_column_names_first_line(\"/data/clusterfs/lag/users/jitame/SENT_CORE/pheno/sent_edges_N29681_resid_mostest.txt\")\n",
    "beta_lang, z_lang, snp_lang = data_munger_MOSTEST_betas(common_all_lang, col_names=col_names_lang, cat=\"edges\")\n",
    "\n",
    "#load CCA data\n",
    "cca_lang = pd.read_csv(\"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/CCA_loadings.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b0bc08dc-b263-47c2-82aa-d88327d43008",
   "metadata": {},
   "outputs": [],
   "source": [
    "#set up data nicely\n",
    "z_lang = z_lang.transpose()\n",
    "cca_lang = cca_lang.transpose()\n",
    "\n",
    "cca_lang.to_csv(\"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/CCA_loadings_T.csv\")\n",
    "\n",
    "#classify edges\n",
    "z_lang[\"Category\"] = [decide_pheno(edge) for edge in z_lang.index.values]\n",
    "cca_lang[\"Category\"] = [decide_pheno(edge) for edge in cca_lang.index.values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "88d9f1bf-608e-447c-8e89-0f4cd05df521",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get t-tests\n",
    "snps_t = ttest_all(z_lang)\n",
    "cca_t = ttest_all(cca_lang)\n",
    "\n",
    "snps_t = snp_lang.set_index(\"SNP\").join(snps_t)\n",
    "\n",
    "cca_t.to_csv(\"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/CCA_ttest.csv\")\n",
    "snps_t.to_csv(\"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/snps_lang_ttest.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "545b07f0-ed3c-4244-a4d6-d2affa643f2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mean_all</th>\n",
       "      <th>Std_all</th>\n",
       "      <th>T_test_1samp_all</th>\n",
       "      <th>Log10P_all</th>\n",
       "      <th>Significance_all</th>\n",
       "      <th>Mean_L</th>\n",
       "      <th>Std_L</th>\n",
       "      <th>Mean_R</th>\n",
       "      <th>Std_R</th>\n",
       "      <th>T_test_LR</th>\n",
       "      <th>Log10P_LR</th>\n",
       "      <th>Signficant_LR</th>\n",
       "      <th>Mean_intra</th>\n",
       "      <th>Std_intra</th>\n",
       "      <th>Mean_inter</th>\n",
       "      <th>Std_inter</th>\n",
       "      <th>T_test_intra_inter</th>\n",
       "      <th>Log10P_intra_inter</th>\n",
       "      <th>Significant_intra_inter</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>score_read</th>\n",
       "      <td>0.052587</td>\n",
       "      <td>0.147355</td>\n",
       "      <td>8.943215</td>\n",
       "      <td>17.376424</td>\n",
       "      <td>True</td>\n",
       "      <td>0.155013</td>\n",
       "      <td>0.13285</td>\n",
       "      <td>0.042903</td>\n",
       "      <td>0.120708</td>\n",
       "      <td>7.700279</td>\n",
       "      <td>12.715747</td>\n",
       "      <td>True</td>\n",
       "      <td>0.098958</td>\n",
       "      <td>0.138751</td>\n",
       "      <td>0.008657</td>\n",
       "      <td>0.141699</td>\n",
       "      <td>8.056898</td>\n",
       "      <td>14.402327</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>score_dyslexia</th>\n",
       "      <td>0.031073</td>\n",
       "      <td>0.131207</td>\n",
       "      <td>5.934905</td>\n",
       "      <td>8.313121</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.025714</td>\n",
       "      <td>0.139464</td>\n",
       "      <td>0.008606</td>\n",
       "      <td>0.120059</td>\n",
       "      <td>-2.299279</td>\n",
       "      <td>1.654296</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.008554</td>\n",
       "      <td>0.131251</td>\n",
       "      <td>0.068615</td>\n",
       "      <td>0.119613</td>\n",
       "      <td>-7.701175</td>\n",
       "      <td>13.277553</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>score_hand</th>\n",
       "      <td>0.068341</td>\n",
       "      <td>0.116232</td>\n",
       "      <td>14.734495</td>\n",
       "      <td>41.694742</td>\n",
       "      <td>True</td>\n",
       "      <td>0.008906</td>\n",
       "      <td>0.098009</td>\n",
       "      <td>0.050331</td>\n",
       "      <td>0.10976</td>\n",
       "      <td>-3.470828</td>\n",
       "      <td>3.226225</td>\n",
       "      <td>True</td>\n",
       "      <td>0.029618</td>\n",
       "      <td>0.106092</td>\n",
       "      <td>0.105025</td>\n",
       "      <td>0.113487</td>\n",
       "      <td>-8.583301</td>\n",
       "      <td>16.139163</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Mean_all   Std_all T_test_1samp_all Log10P_all  \\\n",
       "score_read      0.052587  0.147355         8.943215  17.376424   \n",
       "score_dyslexia  0.031073  0.131207         5.934905   8.313121   \n",
       "score_hand      0.068341  0.116232        14.734495  41.694742   \n",
       "\n",
       "               Significance_all    Mean_L     Std_L    Mean_R     Std_R  \\\n",
       "score_read                 True  0.155013   0.13285  0.042903  0.120708   \n",
       "score_dyslexia             True -0.025714  0.139464  0.008606  0.120059   \n",
       "score_hand                 True  0.008906  0.098009  0.050331   0.10976   \n",
       "\n",
       "               T_test_LR  Log10P_LR Signficant_LR Mean_intra Std_intra  \\\n",
       "score_read      7.700279  12.715747          True   0.098958  0.138751   \n",
       "score_dyslexia -2.299279   1.654296         False  -0.008554  0.131251   \n",
       "score_hand     -3.470828   3.226225          True   0.029618  0.106092   \n",
       "\n",
       "               Mean_inter Std_inter T_test_intra_inter Log10P_intra_inter  \\\n",
       "score_read       0.008657  0.141699           8.056898          14.402327   \n",
       "score_dyslexia   0.068615  0.119613          -7.701175          13.277553   \n",
       "score_hand       0.105025  0.113487          -8.583301          16.139163   \n",
       "\n",
       "               Significant_intra_inter  \n",
       "score_read                        True  \n",
       "score_dyslexia                    True  \n",
       "score_hand                        True  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cca_t.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8742019f-fe78-4831-a199-5f94173feb32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>GP</th>\n",
       "      <th>BP</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Mean_all</th>\n",
       "      <th>Std_all</th>\n",
       "      <th>T_test_1samp_all</th>\n",
       "      <th>Log10P_all</th>\n",
       "      <th>Significance_all</th>\n",
       "      <th>...</th>\n",
       "      <th>T_test_LR</th>\n",
       "      <th>Log10P_LR</th>\n",
       "      <th>Signficant_LR</th>\n",
       "      <th>Mean_intra</th>\n",
       "      <th>Std_intra</th>\n",
       "      <th>Mean_inter</th>\n",
       "      <th>Std_inter</th>\n",
       "      <th>T_test_intra_inter</th>\n",
       "      <th>Log10P_intra_inter</th>\n",
       "      <th>Significant_intra_inter</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs2717066</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>58029645</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.372232</td>\n",
       "      <td>1.454328</td>\n",
       "      <td>6.414026</td>\n",
       "      <td>9.554491</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>1.953681</td>\n",
       "      <td>1.28688</td>\n",
       "      <td>False</td>\n",
       "      <td>0.35788</td>\n",
       "      <td>1.624558</td>\n",
       "      <td>0.385828</td>\n",
       "      <td>1.272069</td>\n",
       "      <td>-0.240525</td>\n",
       "      <td>0.091514</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs17273020</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>17415705</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.304267</td>\n",
       "      <td>1.279956</td>\n",
       "      <td>25.535918</td>\n",
       "      <td>98.464863</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.577837</td>\n",
       "      <td>9.681313</td>\n",
       "      <td>True</td>\n",
       "      <td>1.265073</td>\n",
       "      <td>1.393434</td>\n",
       "      <td>1.341399</td>\n",
       "      <td>1.161054</td>\n",
       "      <td>-0.746646</td>\n",
       "      <td>0.341457</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs35124509</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>89521693</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.734473</td>\n",
       "      <td>2.758324</td>\n",
       "      <td>-6.672832</td>\n",
       "      <td>10.258363</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.91558</td>\n",
       "      <td>0.442959</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.920123</td>\n",
       "      <td>2.947472</td>\n",
       "      <td>-0.558593</td>\n",
       "      <td>2.553885</td>\n",
       "      <td>-1.643908</td>\n",
       "      <td>0.996986</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs143322006</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>134976392</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>-1.077874</td>\n",
       "      <td>1.269232</td>\n",
       "      <td>-21.281713</td>\n",
       "      <td>75.360026</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.332051</td>\n",
       "      <td>4.696946</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.948407</td>\n",
       "      <td>1.326237</td>\n",
       "      <td>-1.200527</td>\n",
       "      <td>1.199942</td>\n",
       "      <td>2.498403</td>\n",
       "      <td>1.89516</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs2279829</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>147106319</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.022422</td>\n",
       "      <td>1.754194</td>\n",
       "      <td>14.606045</td>\n",
       "      <td>41.082441</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>1.311428</td>\n",
       "      <td>0.719643</td>\n",
       "      <td>False</td>\n",
       "      <td>1.209024</td>\n",
       "      <td>1.96129</td>\n",
       "      <td>0.845642</td>\n",
       "      <td>1.511288</td>\n",
       "      <td>2.606577</td>\n",
       "      <td>2.028602</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs2131466</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>28817206</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.277558</td>\n",
       "      <td>1.65689</td>\n",
       "      <td>-4.197969</td>\n",
       "      <td>4.511132</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>-3.410432</td>\n",
       "      <td>3.132958</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.201433</td>\n",
       "      <td>1.739871</td>\n",
       "      <td>-0.349676</td>\n",
       "      <td>1.570848</td>\n",
       "      <td>1.120878</td>\n",
       "      <td>0.580426</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs2274224</th>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>96039597</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1.925297</td>\n",
       "      <td>1.829976</td>\n",
       "      <td>26.365271</td>\n",
       "      <td>102.981673</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>0.686626</td>\n",
       "      <td>0.307292</td>\n",
       "      <td>False</td>\n",
       "      <td>2.264132</td>\n",
       "      <td>1.778745</td>\n",
       "      <td>1.604297</td>\n",
       "      <td>1.819932</td>\n",
       "      <td>4.587804</td>\n",
       "      <td>5.266702</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs4469802</th>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>134309194</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.130344</td>\n",
       "      <td>1.749955</td>\n",
       "      <td>16.186901</td>\n",
       "      <td>48.784863</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>4.136706</td>\n",
       "      <td>4.340585</td>\n",
       "      <td>True</td>\n",
       "      <td>1.451991</td>\n",
       "      <td>1.762838</td>\n",
       "      <td>0.825627</td>\n",
       "      <td>1.681846</td>\n",
       "      <td>4.553116</td>\n",
       "      <td>5.197138</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs10897597</th>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>80057231</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.73096</td>\n",
       "      <td>1.208722</td>\n",
       "      <td>-15.154677</td>\n",
       "      <td>43.715268</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>0.241499</td>\n",
       "      <td>0.091874</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.893585</td>\n",
       "      <td>1.274368</td>\n",
       "      <td>-0.576894</td>\n",
       "      <td>1.121514</td>\n",
       "      <td>-3.307585</td>\n",
       "      <td>3.002246</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs468213</th>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>59071100</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.052054</td>\n",
       "      <td>1.282573</td>\n",
       "      <td>1.017064</td>\n",
       "      <td>0.509319</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.117985</td>\n",
       "      <td>1.456085</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.023962</td>\n",
       "      <td>1.367023</td>\n",
       "      <td>0.124069</td>\n",
       "      <td>1.192623</td>\n",
       "      <td>-1.446905</td>\n",
       "      <td>0.828498</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs4924967</th>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>19172196</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.076701</td>\n",
       "      <td>1.756291</td>\n",
       "      <td>-1.094418</td>\n",
       "      <td>0.561947</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.776409</td>\n",
       "      <td>0.358415</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.374689</td>\n",
       "      <td>1.796109</td>\n",
       "      <td>0.205603</td>\n",
       "      <td>1.669352</td>\n",
       "      <td>-4.192756</td>\n",
       "      <td>4.501324</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs2732687</th>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>44338694</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.807131</td>\n",
       "      <td>1.342391</td>\n",
       "      <td>-15.067624</td>\n",
       "      <td>43.294492</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.332542</td>\n",
       "      <td>0.735938</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.415637</td>\n",
       "      <td>1.337184</td>\n",
       "      <td>-1.178019</td>\n",
       "      <td>1.23793</td>\n",
       "      <td>7.412783</td>\n",
       "      <td>12.395438</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs7234875</th>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>73114340</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.875131</td>\n",
       "      <td>1.382668</td>\n",
       "      <td>-15.861167</td>\n",
       "      <td>47.169723</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>1.221387</td>\n",
       "      <td>0.651917</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.824481</td>\n",
       "      <td>1.492909</td>\n",
       "      <td>-0.923116</td>\n",
       "      <td>1.267564</td>\n",
       "      <td>0.893376</td>\n",
       "      <td>0.429459</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs2360257</th>\n",
       "      <td>XY</td>\n",
       "      <td>0</td>\n",
       "      <td>1896324</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.64435</td>\n",
       "      <td>0.932131</td>\n",
       "      <td>17.323059</td>\n",
       "      <td>54.517198</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.154845</td>\n",
       "      <td>8.625289</td>\n",
       "      <td>True</td>\n",
       "      <td>0.3225</td>\n",
       "      <td>0.957654</td>\n",
       "      <td>0.949261</td>\n",
       "      <td>0.795028</td>\n",
       "      <td>-8.935011</td>\n",
       "      <td>17.346288</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            CHR GP         BP A1 A2  Mean_all   Std_all T_test_1samp_all  \\\n",
       "SNP                                                                        \n",
       "rs2717066     2  0   58029645  A  C  0.372232  1.454328         6.414026   \n",
       "rs17273020    3  0   17415705  T  C  1.304267  1.279956        25.535918   \n",
       "rs35124509    3  0   89521693  C  T -0.734473  2.758324        -6.672832   \n",
       "rs143322006   3  0  134976392  T  C -1.077874  1.269232       -21.281713   \n",
       "rs2279829     3  0  147106319  T  C  1.022422  1.754194        14.606045   \n",
       "rs2131466     4  0   28817206  G  T -0.277558   1.65689        -4.197969   \n",
       "rs2274224    10  0   96039597  C  G  1.925297  1.829976        26.365271   \n",
       "rs4469802    10  0  134309194  A  G  1.130344  1.749955        16.186901   \n",
       "rs10897597   11  0   80057231  A  G  -0.73096  1.208722       -15.154677   \n",
       "rs468213     14  0   59071100  G  C  0.052054  1.282573         1.017064   \n",
       "rs4924967    17  0   19172196  G  T -0.076701  1.756291        -1.094418   \n",
       "rs2732687    17  0   44338694  A  G -0.807131  1.342391       -15.067624   \n",
       "rs7234875    18  0   73114340  C  T -0.875131  1.382668       -15.861167   \n",
       "rs2360257    XY  0    1896324  A  G   0.64435  0.932131        17.323059   \n",
       "\n",
       "             Log10P_all Significance_all  ... T_test_LR Log10P_LR  \\\n",
       "SNP                                       ...                       \n",
       "rs2717066      9.554491             True  ...  1.953681   1.28688   \n",
       "rs17273020    98.464863             True  ... -6.577837  9.681313   \n",
       "rs35124509    10.258363             True  ...  -0.91558  0.442959   \n",
       "rs143322006   75.360026             True  ... -4.332051  4.696946   \n",
       "rs2279829     41.082441             True  ...  1.311428  0.719643   \n",
       "rs2131466      4.511132             True  ... -3.410432  3.132958   \n",
       "rs2274224    102.981673             True  ...  0.686626  0.307292   \n",
       "rs4469802     48.784863             True  ...  4.136706  4.340585   \n",
       "rs10897597    43.715268             True  ...  0.241499  0.091874   \n",
       "rs468213       0.509319            False  ... -2.117985  1.456085   \n",
       "rs4924967      0.561947            False  ...  0.776409  0.358415   \n",
       "rs2732687     43.294492             True  ... -1.332542  0.735938   \n",
       "rs7234875     47.169723             True  ...  1.221387  0.651917   \n",
       "rs2360257     54.517198             True  ... -6.154845  8.625289   \n",
       "\n",
       "            Signficant_LR Mean_intra Std_intra Mean_inter Std_inter  \\\n",
       "SNP                                                                   \n",
       "rs2717066           False    0.35788  1.624558   0.385828  1.272069   \n",
       "rs17273020           True   1.265073  1.393434   1.341399  1.161054   \n",
       "rs35124509          False  -0.920123  2.947472  -0.558593  2.553885   \n",
       "rs143322006          True  -0.948407  1.326237  -1.200527  1.199942   \n",
       "rs2279829           False   1.209024   1.96129   0.845642  1.511288   \n",
       "rs2131466            True  -0.201433  1.739871  -0.349676  1.570848   \n",
       "rs2274224           False   2.264132  1.778745   1.604297  1.819932   \n",
       "rs4469802            True   1.451991  1.762838   0.825627  1.681846   \n",
       "rs10897597          False  -0.893585  1.274368  -0.576894  1.121514   \n",
       "rs468213            False  -0.023962  1.367023   0.124069  1.192623   \n",
       "rs4924967           False  -0.374689  1.796109   0.205603  1.669352   \n",
       "rs2732687           False  -0.415637  1.337184  -1.178019   1.23793   \n",
       "rs7234875           False  -0.824481  1.492909  -0.923116  1.267564   \n",
       "rs2360257            True     0.3225  0.957654   0.949261  0.795028   \n",
       "\n",
       "            T_test_intra_inter Log10P_intra_inter Significant_intra_inter  \n",
       "SNP                                                                        \n",
       "rs2717066            -0.240525           0.091514                   False  \n",
       "rs17273020           -0.746646           0.341457                   False  \n",
       "rs35124509           -1.643908           0.996986                   False  \n",
       "rs143322006           2.498403            1.89516                   False  \n",
       "rs2279829             2.606577           2.028602                   False  \n",
       "rs2131466             1.120878           0.580426                   False  \n",
       "rs2274224             4.587804           5.266702                    True  \n",
       "rs4469802             4.553116           5.197138                    True  \n",
       "rs10897597           -3.307585           3.002246                    True  \n",
       "rs468213             -1.446905           0.828498                   False  \n",
       "rs4924967            -4.192756           4.501324                    True  \n",
       "rs2732687             7.412783          12.395438                    True  \n",
       "rs7234875             0.893376           0.429459                   False  \n",
       "rs2360257            -8.935011          17.346288                    True  \n",
       "\n",
       "[14 rows x 24 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps_t.head(n=14)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0dc8f70-edf9-47f6-b06f-c2c6d643efae",
   "metadata": {},
   "source": [
    "## Hemispheric differences: SNPS + polygenic scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6f5763f3-21a7-4086-a3ab-0f3d7d837eba",
   "metadata": {},
   "outputs": [],
   "source": [
    "common_top_asym = pd.DataFrame.from_dict({\"chr_no\":[3,3,3], \"ID\":[\"rs7635916\", \"rs2279829\", \"rs13321297\"]})\n",
    "col_names_asym = load_column_names_first_line(\"/data/clusterfs/lag/users/jitame/SENT_CORE/pheno/sent_edges_asym_N29681_resid_mostest.txt\")\n",
    "beta_asym, z_asym, snp_asym = data_munger_MOSTEST_betas(common_top_asym, col_names=col_names_asym, cat=\"edges_HD\")\n",
    "\n",
    "cca_asym = pd.read_csv(\"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/CCA_loadings_asym.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0cd91b08-2670-4afd-a7ee-b45795d197f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alpha: 0.016666666666666666\n",
      "[5.16641539e-01 2.24640791e-01 4.31382524e-14]\n",
      "Alpha: 0.016666666666666666\n",
      "[1.01201159e-14 1.04991842e-05 1.03738157e-03]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>GP</th>\n",
       "      <th>BP</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>Mean_all</th>\n",
       "      <th>Std_all</th>\n",
       "      <th>T_test_1samp_all</th>\n",
       "      <th>Log10P_all</th>\n",
       "      <th>Significance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs7635916</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>89633140</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.111436</td>\n",
       "      <td>1.729337</td>\n",
       "      <td>-0.650796</td>\n",
       "      <td>0.286811</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs2279829</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>147106319</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.182701</td>\n",
       "      <td>1.510362</td>\n",
       "      <td>1.22169</td>\n",
       "      <td>0.648511</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs13321297</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>17528764</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>-1.084681</td>\n",
       "      <td>1.249534</td>\n",
       "      <td>-8.767063</td>\n",
       "      <td>13.365137</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           CHR GP         BP A1 A2  Mean_all   Std_all T_test_1samp_all  \\\n",
       "SNP                                                                       \n",
       "rs7635916    3  0   89633140  A  G -0.111436  1.729337        -0.650796   \n",
       "rs2279829    3  0  147106319  T  C  0.182701  1.510362          1.22169   \n",
       "rs13321297   3  0   17528764  A  T -1.084681  1.249534        -8.767063   \n",
       "\n",
       "           Log10P_all Significance  \n",
       "SNP                                 \n",
       "rs7635916    0.286811        False  \n",
       "rs2279829    0.648511        False  \n",
       "rs13321297  13.365137         True  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#turn data\n",
    "z_asym = z_asym.transpose()\n",
    "cca_asym = cca_asym.transpose()\n",
    "\n",
    "#do t-testing\n",
    "t_asym = ttest_asym(z_asym)\n",
    "t_asym = snp_asym.set_index(\"SNP\").join(t_asym)\n",
    "CCA_t_asym = ttest_asym(cca_asym)\n",
    "\n",
    "#save\n",
    "t_asym.to_csv(\"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/snps_t_asym.csv\")\n",
    "CCA_t_asym.to_csv(\"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/cca_t_asym.csv\")\n",
    "\n",
    "t_asym.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "22fc1b5a-b607-4fd1-8cdb-b301ddd9dcca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mean_all</th>\n",
       "      <th>Std_all</th>\n",
       "      <th>T_test_1samp_all</th>\n",
       "      <th>Log10P_all</th>\n",
       "      <th>Significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>score_read</th>\n",
       "      <td>0.16929</td>\n",
       "      <td>0.188859</td>\n",
       "      <td>9.053001</td>\n",
       "      <td>13.994815</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>score_dyslexia</th>\n",
       "      <td>-0.068732</td>\n",
       "      <td>0.149694</td>\n",
       "      <td>-4.637187</td>\n",
       "      <td>4.978844</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>score_hand</th>\n",
       "      <td>-0.053669</td>\n",
       "      <td>0.160493</td>\n",
       "      <td>-3.377257</td>\n",
       "      <td>2.984061</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Mean_all   Std_all T_test_1samp_all Log10P_all Significance\n",
       "score_read       0.16929  0.188859         9.053001  13.994815         True\n",
       "score_dyslexia -0.068732  0.149694        -4.637187   4.978844         True\n",
       "score_hand     -0.053669  0.160493        -3.377257   2.984061         True"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CCA_t_asym.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06570f55-56ae-496e-84d2-c63b70b048fc",
   "metadata": {},
   "source": [
    "## Rare variant burden scores: language network + HDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "34f3ba17-8491-4506-9f04-def2e90901f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "rare_genes = pd.read_csv(os.path.join(\"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/regenie_gene_based/\", \"all_exome_results.csv\"), engine=\"pyarrow\")\n",
    "rare_genes = add_gene_names(rare_genes)\n",
    "\n",
    "data_burden_strict = rare_genes[np.array([rare_genes[\"TEST\"] == \"ADD\", rare_genes[\"ALLELE1\"] == \"Strict.0.01\"]).all(0).reshape((len(rare_genes), 1))]\n",
    "data_burden_broad = rare_genes[np.array([rare_genes[\"TEST\"] == \"ADD\", rare_genes[\"ALLELE1\"] == \"Broad.0.01\"]).all(0).reshape((len(rare_genes), 1))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "74ae7014-8d28-4a7d-933a-e11f58d9db23",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes=[\"MANEAL\", \"DUSP29\", \"SLC25A48\", \"TRIP11\", \"NIBAN1\"]\n",
    "\n",
    "heritable_lang = load_column_names(\"/data/clusterfs/lag/users/jitame/SENT_CORE/heritable_edges.txt\")\n",
    "heritable_b_lang = [\"BETA_sent_edges_\"+x for x in heritable_lang]\n",
    "heritable_se_lang = [\"SE_sent_edges_\"+x for x in heritable_lang]\n",
    "lang_burden_broad_sig = data_burden_broad.set_index(\"gene_name\").loc[genes, :]\n",
    "\n",
    "beta_df_broad = lang_burden_broad_sig[heritable_b_lang]\n",
    "beta_df_broad.columns=[x[16:] for x in beta_df_broad.columns]\n",
    "\n",
    "se_df_broad = lang_burden_broad_sig[heritable_se_lang]\n",
    "se_df_broad.columns=[x[16:] for x in se_df_broad.columns]\n",
    "\n",
    "z_np = beta_df_broad.to_numpy() / se_df_broad.to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "72c8e111-5199-4f2b-ba21-7ee5d299e280",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = [\"WDCP\", \"DDX25\"]\n",
    "\n",
    "heritable_asym = load_column_names(\"/data/clusterfs/lag/users/jitame/SENT_CORE/heritable_edges_asym.txt\")\n",
    "heritable_b_asym = [\"BETA_sent_edges_\"+x for x in heritable_asym]\n",
    "heritable_se_asym = [\"SE_sent_edges_\"+x for x in heritable_asym]\n",
    "asym_strict_sig = data_burden_strict.set_index(\"gene_name\").loc[genes, :]\n",
    "\n",
    "beta_df_strict = asym_strict_sig[heritable_b_asym]\n",
    "beta_df_strict.columns=[x[16:] for x in beta_df_strict.columns]\n",
    "\n",
    "se_df_strict = asym_strict_sig[heritable_se_asym]\n",
    "se_df_strict.columns=[x[16:] for x in se_df_strict.columns]\n",
    "\n",
    "z_np_asym = beta_df_strict.to_numpy() / se_df_strict.to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "14a5fd98-16d4-4326-becf-cfac0177e573",
   "metadata": {},
   "outputs": [],
   "source": [
    "z_df = pd.DataFrame(z_np, columns=beta_df_broad.columns, index=beta_df_broad.index).T\n",
    "z_df_asym = pd.DataFrame(z_np_asym, columns=beta_df_strict.columns, index=beta_df_strict.index).T\n",
    "\n",
    "z_df[\"Category\"] = [decide_pheno(edge) for edge in z_df.index.values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d62b7364-a5b6-447a-b358-61a166438e38",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alpha: 0.025\n",
      "[4.27917297e-05 8.45808673e-21]\n"
     ]
    }
   ],
   "source": [
    "rare_t = ttest_all(z_df)\n",
    "rare_t_asym = ttest_asym(z_df_asym)\n",
    "\n",
    "rare_t.to_csv(\"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/rare_t_lang.csv\")\n",
    "rare_t_asym.to_csv(\"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/rare_t_asym.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2f9349ba-fb6e-4699-a2ea-71b153e94691",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mean_all</th>\n",
       "      <th>Std_all</th>\n",
       "      <th>T_test_1samp_all</th>\n",
       "      <th>Log10P_all</th>\n",
       "      <th>Significance_all</th>\n",
       "      <th>Mean_L</th>\n",
       "      <th>Std_L</th>\n",
       "      <th>Mean_R</th>\n",
       "      <th>Std_R</th>\n",
       "      <th>T_test_LR</th>\n",
       "      <th>Log10P_LR</th>\n",
       "      <th>Signficant_LR</th>\n",
       "      <th>Mean_intra</th>\n",
       "      <th>Std_intra</th>\n",
       "      <th>Mean_inter</th>\n",
       "      <th>Std_inter</th>\n",
       "      <th>T_test_intra_inter</th>\n",
       "      <th>Log10P_intra_inter</th>\n",
       "      <th>Significant_intra_inter</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MANEAL</th>\n",
       "      <td>-1.113141</td>\n",
       "      <td>0.884362</td>\n",
       "      <td>-31.542788</td>\n",
       "      <td>130.867668</td>\n",
       "      <td>True</td>\n",
       "      <td>-1.144357</td>\n",
       "      <td>0.752854</td>\n",
       "      <td>-1.00325</td>\n",
       "      <td>0.995739</td>\n",
       "      <td>-1.393627</td>\n",
       "      <td>0.78397</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.073803</td>\n",
       "      <td>0.885506</td>\n",
       "      <td>-1.150408</td>\n",
       "      <td>0.881659</td>\n",
       "      <td>1.085126</td>\n",
       "      <td>0.555514</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DUSP29</th>\n",
       "      <td>-0.761958</td>\n",
       "      <td>1.039243</td>\n",
       "      <td>-18.373575</td>\n",
       "      <td>59.935635</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.835673</td>\n",
       "      <td>0.890321</td>\n",
       "      <td>-0.793821</td>\n",
       "      <td>1.136983</td>\n",
       "      <td>-0.357308</td>\n",
       "      <td>0.141999</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.814747</td>\n",
       "      <td>1.021342</td>\n",
       "      <td>-0.711948</td>\n",
       "      <td>1.053485</td>\n",
       "      <td>-1.239505</td>\n",
       "      <td>0.666306</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC25A48</th>\n",
       "      <td>0.800083</td>\n",
       "      <td>1.08167</td>\n",
       "      <td>18.536168</td>\n",
       "      <td>60.783021</td>\n",
       "      <td>True</td>\n",
       "      <td>0.663567</td>\n",
       "      <td>1.156638</td>\n",
       "      <td>0.907603</td>\n",
       "      <td>1.097782</td>\n",
       "      <td>-1.886722</td>\n",
       "      <td>1.22076</td>\n",
       "      <td>False</td>\n",
       "      <td>0.785585</td>\n",
       "      <td>1.134176</td>\n",
       "      <td>0.813817</td>\n",
       "      <td>1.029271</td>\n",
       "      <td>-0.326684</td>\n",
       "      <td>0.128418</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRIP11</th>\n",
       "      <td>1.131913</td>\n",
       "      <td>0.971633</td>\n",
       "      <td>29.193812</td>\n",
       "      <td>118.306362</td>\n",
       "      <td>True</td>\n",
       "      <td>1.621236</td>\n",
       "      <td>1.022125</td>\n",
       "      <td>0.80931</td>\n",
       "      <td>0.877088</td>\n",
       "      <td>7.432189</td>\n",
       "      <td>11.961986</td>\n",
       "      <td>True</td>\n",
       "      <td>1.215273</td>\n",
       "      <td>1.035286</td>\n",
       "      <td>1.052941</td>\n",
       "      <td>0.900126</td>\n",
       "      <td>2.098303</td>\n",
       "      <td>1.440358</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NIBAN1</th>\n",
       "      <td>0.818342</td>\n",
       "      <td>1.057588</td>\n",
       "      <td>19.390913</td>\n",
       "      <td>65.271162</td>\n",
       "      <td>True</td>\n",
       "      <td>0.400703</td>\n",
       "      <td>0.938554</td>\n",
       "      <td>1.407184</td>\n",
       "      <td>1.081924</td>\n",
       "      <td>-8.663581</td>\n",
       "      <td>15.560182</td>\n",
       "      <td>True</td>\n",
       "      <td>0.903943</td>\n",
       "      <td>1.130917</td>\n",
       "      <td>0.737246</td>\n",
       "      <td>0.976185</td>\n",
       "      <td>1.97884</td>\n",
       "      <td>1.316313</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Mean_all   Std_all T_test_1samp_all  Log10P_all Significance_all  \\\n",
       "MANEAL   -1.113141  0.884362       -31.542788  130.867668             True   \n",
       "DUSP29   -0.761958  1.039243       -18.373575   59.935635             True   \n",
       "SLC25A48  0.800083   1.08167        18.536168   60.783021             True   \n",
       "TRIP11    1.131913  0.971633        29.193812  118.306362             True   \n",
       "NIBAN1    0.818342  1.057588        19.390913   65.271162             True   \n",
       "\n",
       "            Mean_L     Std_L    Mean_R     Std_R T_test_LR  Log10P_LR  \\\n",
       "MANEAL   -1.144357  0.752854  -1.00325  0.995739 -1.393627    0.78397   \n",
       "DUSP29   -0.835673  0.890321 -0.793821  1.136983 -0.357308   0.141999   \n",
       "SLC25A48  0.663567  1.156638  0.907603  1.097782 -1.886722    1.22076   \n",
       "TRIP11    1.621236  1.022125   0.80931  0.877088  7.432189  11.961986   \n",
       "NIBAN1    0.400703  0.938554  1.407184  1.081924 -8.663581  15.560182   \n",
       "\n",
       "         Signficant_LR Mean_intra Std_intra Mean_inter Std_inter  \\\n",
       "MANEAL           False  -1.073803  0.885506  -1.150408  0.881659   \n",
       "DUSP29           False  -0.814747  1.021342  -0.711948  1.053485   \n",
       "SLC25A48         False   0.785585  1.134176   0.813817  1.029271   \n",
       "TRIP11            True   1.215273  1.035286   1.052941  0.900126   \n",
       "NIBAN1            True   0.903943  1.130917   0.737246  0.976185   \n",
       "\n",
       "         T_test_intra_inter Log10P_intra_inter Significant_intra_inter  \n",
       "MANEAL             1.085126           0.555514                   False  \n",
       "DUSP29            -1.239505           0.666306                   False  \n",
       "SLC25A48          -0.326684           0.128418                   False  \n",
       "TRIP11             2.098303           1.440358                   False  \n",
       "NIBAN1              1.97884           1.316313                   False  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rare_t.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "dab7d11e-c976-4742-b90f-c959b862787a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mean_all</th>\n",
       "      <th>Std_all</th>\n",
       "      <th>T_test_1samp_all</th>\n",
       "      <th>Log10P_all</th>\n",
       "      <th>Significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WDCP</th>\n",
       "      <td>-0.491599</td>\n",
       "      <td>1.160875</td>\n",
       "      <td>-4.276869</td>\n",
       "      <td>4.36864</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DDX25</th>\n",
       "      <td>-1.54418</td>\n",
       "      <td>1.320657</td>\n",
       "      <td>-11.808861</td>\n",
       "      <td>20.072728</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Mean_all   Std_all T_test_1samp_all Log10P_all Significance\n",
       "WDCP  -0.491599  1.160875        -4.276869    4.36864         True\n",
       "DDX25  -1.54418  1.320657       -11.808861  20.072728         True"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rare_t_asym.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a5e160cd-7040-465e-b85a-bc926632a179",
   "metadata": {},
   "outputs": [],
   "source": [
    "z_df.to_csv(\"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/rare_z_lang.csv\")\n",
    "z_df_asym.to_csv(\"/data/clusterfs/lag/users/jitame/SENT_CORE/geno/rare_z_asym.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:results_env]",
   "language": "python",
   "name": "conda-env-results_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
